python-package/lightgbm/basic.py

```diff
@@ -2081,16 +2081,39 @@ def __init__(self, params=None, train_set=None, model_file=None, model_str=None,
                 if alias in params:
                     machines = params[alias]
                     if isinstance(machines, str):
-                        num_machines = len(machines.split(','))
+                        num_machines_from_machine_list = len(machines.split(','))
                     elif isinstance(machines, (list, set)):
-                        num_machines = len(machines)
+                        num_machines_from_machine_list = len(machines)
                         machines = ','.join(machines)
                     else:
                         raise ValueError("Invalid machines in params.")
-                    self.set_network(machines,
-                                     local_listen_port=params.get("local_listen_port", 12400),
-                                     listen_time_out=params.get("listen_time_out", 120),
-                                     num_machines=params.setdefault("num_machines", num_machines))
+
+                    local_listen_port = 12400
+                    for port_param in _ConfigAliases.get("local_listen_port"):
+                        val = params.get(port_param)
+                        if val is not None:
+                            local_listen_port = val
+                            params["local_listen_port"] = val
+                            break
+
+                    # if num_machines is in params, use that. Otherwise, take
+                    # num_machines implied by "machines"
+                    num_machines_from_params = None
+                    for machine_param in _ConfigAliases.get("num_machines"):
+                        val = params.pop(machine_param)
+                        if num_machines_from_params is None:
+                            num_machines_from_params = val
+                            break
```

Comment:
Remove all aliases from `params` to keep `params` clean like here
https://github.com/microsoft/LightGBM/blob/d9a96c90cb479cef87047ba20517d97982b563eb/python-package/lightgbm/engine.py#L146-L150
```suggestion
                    for port_param in _ConfigAliases.get("local_listen_port"):
                        val = params.get(port_param)
                        if val is not None:
                            local_listen_port = val
                            params["local_listen_port"] = val

                    # if num_machines is in params, use that. Otherwise, take
                    # num_machines implied by "machines"
                    num_machines_from_params = None
                    for machine_param in _ConfigAliases.get("num_machines"):
                        val = params.pop(machine_param)
                        if num_machines_from_params is None:
                            num_machines_from_params = val
```

Reply:
@StrikerRUS since this pattern is repeated in a few places, what do you think about me changing this PR to introduce a function that does this?

Like this

```python
def _choose_param_value(main_param_name: str, params: Dict[str, Any], default_value: Any):
    # avoid side effects on passed-in parameters
    params = deepcopy(params)

    # find a value, and remove other aliases with .pop()
    found_value = None
    for param in _ConfigAliases.get(main_param_name):
        val = params.pop(param)
        if found_value is None and val is not None:
            found_value = val

    if found_value is not None:
        params[main_param_name] = found_value
    else:
        params[main_param_name] = default_value

    return params


params = _choose_param_value(
    main_param_name="local_listen_port",
    params=params,
    default_value=12400
)
params = _choose_param_value(
    main_param_name="num_machines",
    params=params,
    default_value=num_machines_from_machine_list
)
```

Then we could re-use this everywhere that we want to do this type of operation. That would allow us to test it independently and to add logging in one place if we wanted (like a warning about duplicated parameters).

Reply:
Пкуфе швуфб Ш дшлув ше!

Reply:
thanks! Adding this right now. Ignore the diff here for now, I'll re-request a review when this is ready

Reply:
ok @StrikerRUS I think this is ready for another review.

I'm also proposing moving all of the parameter stuff in `lightgbm.dask._train()` up near the beginning of that function. That will avoid the case where you have to wait for LightGBM to do all the data setup stuff, only to find out afterwards that training fails because of an issue in parameters.